tests/chainer_tests/links_tests/connection_tests/test_peephole.py

```diff
@@ -0,0 +1,211 @@
+    y = o * xp.tanh(c_next)
+    return c_next, y
+
+
+@testing.parameterize(
+    {'in_size': 10, 'out_size': 10},
+    {'in_size': 10, 'out_size': 40},
+)
+class TestPeephole(unittest.TestCase):
+
+    def setUp(self):
+        self.link = links.Peephole(self.in_size, self.out_size)
+        upward = self.link.upward.W.data
+        upward[...] = numpy.random.uniform(-1, 1, upward.shape)
+        lateral = self.link.lateral.W.data
+        lateral[...] = numpy.random.uniform(-1, 1, lateral.shape)
+        peep_i = self.link.peep_i.W.data
+        peep_i[...] = numpy.random.uniform(-1, 1, peep_i.shape)
+        peep_f = self.link.peep_f.W.data
+        peep_f[...] = numpy.random.uniform(-1, 1, peep_f.shape)
+        peep_o = self.link.peep_o.W.data
+        peep_o[...] = numpy.random.uniform(-1, 1, peep_o.shape)
+
+        self.upward = upward.copy()
+        self.lateral = lateral.copy()
+        self.peep_i = peep_i.copy()
+        self.peep_f = peep_f.copy()
+        self.peep_o = peep_o.copy()
+
+        c_shape = (1, self.out_size)
+        h_shape = (1, self.out_size)
+        x_shape = (4, self.in_size)
+        gy_shape = (4, self.out_size)
+        self.c = numpy.zeros(c_shape).astype(numpy.float32)
+        self.h = numpy.zeros(h_shape).astype(numpy.float32)
+        self.x = numpy.random.uniform(-1, 1, x_shape).astype(numpy.float32)
+        self.gy = numpy.random.uniform(-1, 1, gy_shape).astype(numpy.float32)
+
+    def _forward(self, link, x):
+        return link(x)
+
+    def check_forward(self, c_data, h_data, x_data):
+        x = chainer.Variable(x_data)
+
+        h1 = self.link(x)
+        c1_expect, h1_expect = _peephole(self.link, c_data, h_data, x_data)
+        gradient_check.assert_allclose(h1.data, h1_expect)
+        gradient_check.assert_allclose(self.link.c.data, c1_expect)
+        gradient_check.assert_allclose(self.link.h.data, h1_expect)
+
+        h2 = self.link(x)
+        c2_expect, h2_expect = _peephole(self.link,
+                                         c1_expect, h1_expect, x_data)
+        gradient_check.assert_allclose(h2.data, h2_expect)
+        gradient_check.assert_allclose(self.link.c.data, c2_expect)
+        gradient_check.assert_allclose(self.link.h.data, h2_expect)
+
+    def test_forward_cpu(self):
+        self.check_forward(self.c, self.h, self.x)
+
+    @attr.gpu
+    def test_forward_gpu(self):
+        self.link.to_gpu()
+        self.check_forward(cuda.to_gpu(self.c),
+                           cuda.to_gpu(self.h),
+                           cuda.to_gpu(self.x))
+
+    def check_backward(self, c_data, h_data, x_data, y_grad):
+        x = chainer.Variable(x_data)
+        y = self._forward(self.link, x)
+        y.grad = y_grad
+        y.backward()
+
+        def f():
+            c, y = _peephole(self.link, c_data, h_data, x_data)
+            return y,
+        gx, = gradient_check.numerical_grad(f, (x.data,), (y.grad,))
+        gradient_check.assert_allclose(gx, x.grad, atol=1e-3)
```

Comment:
I understand the meaning of `gradient_check.check_backward` . Then, I do not change the current code.